contracts/DissentVoting.sol

```diff
@@ -0,0 +1,476 @@
+        voteExists(_voteId)
+        returns (
+            bool open,
+            bool executed,
+            uint64 startDate,
+            uint64 snapshotBlock,
+            uint64 supportRequired,
+            uint64 minAcceptQuorum,
+            uint256 yea,
+            uint256 nay,
+            uint256 votingPower,
+            bytes script
+        )
+    {
+        Vote storage vote_ = votes[_voteId];
+
+        open = _isVoteOpen(vote_);
+        executed = vote_.executed;
+        startDate = vote_.startDate;
+        snapshotBlock = vote_.snapshotBlock;
+        supportRequired = vote_.supportRequiredPct;
+        minAcceptQuorum = vote_.minAcceptQuorumPct;
+        yea = vote_.yea;
+        nay = vote_.nay;
+        votingPower = vote_.votingPower;
+        script = vote_.executionScript;
+    }
+
+    /**
+    * @dev Return the state of a voter for a given vote by its ID
+    * @param _voteId Vote identifier
+    * @return VoterState of the requested voter for a certain vote
+    */
+    function getVoterState(uint256 _voteId, address _voter) public view voteExists(_voteId) returns (VoterState) {
+        return votes[_voteId].voters[_voter];
+    }
+
+    // Internal fns
+
+    /**
+    * @dev Internal function to create a new vote
+    * @return voteId id for newly created vote
+    */
+    function _newVote(bytes _executionScript, string _metadata, bool _castVote, bool _executesIfDecided) internal returns (uint256 voteId) {
+        uint64 snapshotBlock = getBlockNumber64() - 1; // avoid double voting in this very block
+        uint256 votingPower = token.totalSupplyAt(snapshotBlock);
+        require(votingPower > 0, ERROR_NO_VOTING_POWER);
+
+        voteId = votesLength++;
+
+        Vote storage vote_ = votes[voteId];
+        vote_.startDate = getTimestamp64();
+        vote_.snapshotBlock = snapshotBlock;
+        vote_.supportRequiredPct = supportRequiredPct;
+        vote_.minAcceptQuorumPct = minAcceptQuorumPct;
+        vote_.votingPower = votingPower;
+        vote_.executionScript = _executionScript;
+
+        emit StartVote(voteId, msg.sender, _metadata);
+
+        if (_castVote && _canVote(voteId, msg.sender)) {
+            _vote(voteId, true, msg.sender, _executesIfDecided);
+        }
+    }
+
+    /**
+    * @dev Internal function to cast a vote. It assumes the queried vote exists.
+    */
+    function _vote(
+        uint256 _voteId,
+        bool _supports,
+        address _voter,
+        bool _executesIfDecided
+    ) internal
+    {
+        Vote storage vote_ = votes[_voteId];
+
+        // This could re-enter, though we can assume the governance token is not malicious
+        uint256 voterStake = token.balanceOfAt(_voter, vote_.snapshotBlock);
+
+        //Removed resetting of vote stake, that was in the original voting-app, as voters can only vote once.
+        if (_supports) {
+            vote_.yea = vote_.yea.add(voterStake);
+            lastYeaVoteTime[_voter] = getTimestamp64();
```

Comment:
We should use the time the proposal/vote that is being voted on was created rather than the current time. This is slightly more complicated and we will need to first make sure that the time is greater than the current time in the mapping before updating... but it prevents an issue when using the dissent oracle with redemptions where a voter would have a very small amount of time to exit due to the timing of their vote relative to a new proposal being created that would cause them to exit.